### PR TITLE
Refactor/refactor build config

### PR DESCRIPTION
**Pull request by xizhao**
1. Pull `config` to its own package
2. Move `Builder` to the `build` package

### CODE REVIEW

File: build/archive.go [hunk 1]
```diff
@@ -11,9 +11,9 @@ import (
 	"strings"
 
 	"github.com/bmatcuk/doublestar"
+	"github.com/fossas/fossa-cli/config"
```
Comment (ilikebits) [inline build/archive.go]: Internal imports should go in a separate section.

File: build/builder.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	// Build runs a best-effort attempt at building the module.
+	Build(m module.Module, force bool) error
+	// Analyze returns the dependencies of a module.
+	Analyze(m module.Module, allowUnresolved bool) ([]module.Dependency, error)
+
+	// IsBuilt checks whether a module has been built.
+	IsBuilt(m module.Module, allowUnresolved bool) (bool, error)
+
+	// IsModule takes a configKey and returns whether it is valid to elect this Builder.
+	IsModule(configKey string) (bool, error)
```
Comment (ilikebits) [inline build/builder.go]: If you're not planning to use `IsModule` in `init`, remove it from the interface and remove all the implementations.

File: cmd/fossa/analyze.go [hunk 3]
```diff
@@ -38,39 +34,39 @@ func analyzeCmd(c *cli.Context) {
...
 type analysis map[analysisKey][]module.Dependency
 
-func doAnalyze(modules []moduleConfig, allowUnresolved bool) (analysis, error) {
+func doAnalyze(modules []config.ModuleConfig, allowUnresolved bool) (analysis, error) {
 	analysisLogger.Debugf("Running analysis on modules: %#v", modules)
 	dependencies := make(analysis)
 
-	for _, moduleConfig := range modules {
-		builder, m, err := resolveModuleConfig(moduleConfig)
+	for _, ModuleConfig := range modules {
```
Comment (ilikebits) [inline cmd/fossa/analyze.go]: Why is this now capitalised?

File: cmd/fossa/build.go [hunk 4]
```diff
@@ -4,27 +4,24 @@ import (
 	"fmt"
 	"os"
 
+	config "github.com/fossas/fossa-cli/config"
```
Comment (ilikebits) [inline cmd/fossa/build.go]: Internal imports should go in another section.

File: cmd/fossa/build.go [hunk 5]
```diff
@@ -4,27 +4,24 @@ import (
...
 		buildLogger.Fatalf("Could not load configuration: %s", err.Error())
 	}
-	if len(config.modules) == 0 {
+	if len(conf.Modules) == 0 {
 		buildLogger.Fatal("No modules specified.")
 	}
 
-	for _, moduleConfig := range config.modules {
-		builder, module, err := resolveModuleConfig(moduleConfig)
+	for _, ModuleConfig := range conf.Modules {
```
Comment (ilikebits) [inline cmd/fossa/build.go]: Why is this now capitalised?

File: cmd/fossa/config.go [hunk 6]
```diff
@@ -1,263 +0,0 @@
-package main
```
Comment (ilikebits) [inline cmd/fossa/config.go]: You should use `git mv` to preserve the movement of a file in history.

File: cmd/fossa/init.go [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package main
```
Comment (ilikebits) [inline cmd/fossa/init.go]: I'd take this out of this PR.

File: cmd/fossa/main.go [hunk 8]
```diff
@@ -143,12 +152,12 @@ func setupModule(config moduleConfig, manifestName string, moduleType module.Typ
 	return m, nil
 }
 
-func resolveModuleConfig(moduleConfig moduleConfig) (module.Builder, module.Module, error) {
-	mainLogger.Debugf("Resolving moduleConfig: %#v", moduleConfig)
+func resolveModuleConfig(moduleConfig config.ModuleConfig) (build.Builder, module.Module, error) {
+	mainLogger.Debugf("Resolving ModuleConfig: %#v", moduleConfig)
```
Comment (ilikebits) [inline cmd/fossa/main.go]: Why is this now capitalised?

File: cmd/fossa/test.go [hunk 9]
```diff
@@ -9,6 +9,7 @@ import (
 	"time"
 
 	"github.com/briandowns/spinner"
+	config "github.com/fossas/fossa-cli/config"
```
Comment (ilikebits) [inline cmd/fossa/test.go]: Internal imports should go in their own section.

File: cmd/fossa/main.go [hunk 10]
```diff
@@ -234,17 +243,17 @@ func resolveModuleConfig(moduleConfig moduleConfig) (module.Builder, module.Modu
 		return builder, m, fmt.Errorf("unknown module type: %s", moduleConfig.Type)
 	}
 
-	mainLogger.Debugf("Resolved moduleConfig to: %#v, %#v", builder, m)
+	mainLogger.Debugf("Resolved ModuleConfig to: %#v, %#v", builder, m)
```
Comment (ilikebits) [inline cmd/fossa/main.go]: Again here.

File: cmd/fossa/upload.go [hunk 11]
```diff
@@ -12,15 +12,13 @@ import (
 	logging "github.com/op/go-logging"
 	"github.com/urfave/cli"
 
+	"github.com/fossas/fossa-cli/build"
+	config "github.com/fossas/fossa-cli/config"
```
Comment (ilikebits) [inline cmd/fossa/upload.go]: Why is this import named?

File: config/config.go [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+	if err == nil {
+		return ParseConfigFile(".fossa.yaml")
+	}
+
+	return setDefaultValues(configFileV1{})
+}
+
+// WriteConfigFile writes a config state to yaml
+func WriteConfigFile(filename string, conf configFileV1) error {
+	return errors.New("Not Implemented")
```
Comment (ilikebits) [inline config/config.go]: This should be more descriptive and follow the formatting conventions for error messages. A better message would be `"WriteConfigFile is not implemented"`

File: config/config.go [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+	if err != nil {
+		return config, err
+	}
+
+	return config, nil
+}
+
+// SerializeConfigFile turns a config state to yaml
+func SerializeConfigFile(configFileV1) (string, error) {
+	return "", errors.New("Not Implemented")
```
Comment (ilikebits) [inline config/config.go]: Same error message stuff here.

File: config/config.go [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+			Path: sections[1],
+			Type: ModuleType(sections[0]),
+		})
+	}
+
+	return config
+}
+
+// Initialize creates a CliConfig from cli context
+func Initialize(c *cli.Context) (CliConfig, error) {
```
Comment (ilikebits) [inline config/config.go]: Conventionally, this would be called `func New`

File: build/builder.go (hunk 2 above)
Comment (xizhao) [reply build/builder.go]: I have plans for IsModule in `init`.

File: config/config.go (hunk 13 above)
Comment (xizhao) [reply config/config.go]: This will be removed when `fossa init` is implemented

File: config/config.go (hunk 12 above)
Comment (xizhao) [reply config/config.go]: This will be removed when fossa init is implemented

File: cmd/fossa/upload.go (hunk 11 above)
Comment (xizhao) [reply cmd/fossa/upload.go]: idk

**Comment by xizhao** [comment]
Thanks for comments, will resolve in #49

**Pull request closed without merging**